### ISSUE TITLE
Lint against `super` (and `self`) imports

### ISSUE THREAD

**Issue by vi013t** [rust-lang/rust-clippy#12796]
### What it does

For a lot of crates, it can be convenient and easier to read global crate-based imports (i.e. use `crate::a::b::c` instead of `super::c`, and IDEs can often automatically put in super-style imports, which can then be confusing when also mixed with absolute (`crate::`) imports. It'd be really nice to have a lint that could warn against `super::` style imports, or at least lint against inconsistency between the two (i.e. using both `crate::` and `super::` imports in the same file), or *at least* for using super- and crate- style imports to import different things from the same file. For example, in this module structure:

src/main.rs (defines a `static FOO` and a `static BAR`)
src/directory/mod.rs (needs to import `FOO` and `BAR`)

`src/directory/mod.rs` can do `use crate::FOO` and `use super::BAR`, and it's totally unclear that these are coming from the same place, which can be confusing.

The same goes for `self::` imports; it'd be really convenient to at least be able to warn when different imports import things from the same file, if not warn against import styles in general (I, for example, prefer to only use absolute `crate::` imports in all my files, so it'd be really handy to be able to lint against `super::` and `self::` imports entirely, though I can understand the argument for this being a bit too restrictive).

### Advantage

The new code unifies the imports into a single unified style and makes it clear when imports are coming from the same location. Similarly, "close together" files are often close together in the imports, which isn't necessarily true with a split import style. Also, the imports are often more concise, i.e., doing `use crate::{FOO, BAR}` is less characters than `use crate::FOO; use super::BAR;`. It also would help for situations where IDEs auto-import in a style against what a project is using, which can quickly create messy imports in files. 

### Drawbacks

Some people may prefer to mix import styles, especially for test modules and whatnot (though they could just not enable the lint of course, but obviously that applies to every lint). Some people may have some specific system or preference that involves mixing `crate::`, `super::` and `self::` imports.

### Example

In a file structure as follows:

src/main.rs (defines a `static FOO` and a `static BAR`)
src/directory/mod.rs (needs to import `FOO` and `BAR`)

This code in `src/directory/mod.rs`:

```rust
use crate::FOO;
use super::BAR;
```

Could be written as:

```rust
use crate::{FOO, BAR};
```

**Comment by J-ZhengLi** [comment]
Great idea! but I think maybe it'd be better if it could be implemented in `rustfmt`? Plus it already has a bunch of options to change the style of imports.

**Comment by vi013t** [comment]
Fair point, but I personally think it'd be better living here (or both); It'd make it available for projects not using `rustfmt`, and as you said, `rustfmt` already has a decent amount of complexity and options for customizing imports; I think it would add a lot of complexity to the configuration to allow all different styles (prefer `crate::`, prefer `super::`, prefer `self::`, allow mixing all of these, don't allow mixing any of these, allow mixing only some two of those three, etc.). 

I think as a lint (or a few) it could be quite simple &mdash; It could be lints like `crate_imports = "allow"`, `self_imports = "allow"`, and `"super_imports = "allow"`, which would allow many different combinations based on preference from just those few lints. Alternatively, if it's too big of a jump to add in 3 lints to solve one problem, there could simply be a `mixed_imports = "warn"` lint or something along those lines, that has no preference for `crate::`, `self::` and `super::` imports, but at least warns against mixing those different types. 

Not that this would be *bad* for `rustfmt`, but I think from a configuration standpoint, having it as a lint(s) would be simpler to handle and customize, in addition to giving access to this for projects that don't use `rustfmt`.
